cpp/src/reshape/explode.cu

```diff
@@ -0,0 +1,140 @@
+/*
+ * Copyright (c) 2021, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include <cudf/column/column_device_view.cuh>
+#include <cudf/detail/gather.cuh>
+#include <cudf/detail/nvtx/ranges.hpp>
+#include <cudf/lists/lists_column_view.hpp>
+#include <cudf/reshape.hpp>
+#include <cudf/table/table.hpp>
+#include <cudf/utilities/type_dispatcher.hpp>
+
+#include <rmm/cuda_stream_view.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <thrust/binary_search.h>
+
+#include <memory>
+#include <type_traits>
+#include "thrust/iterator/counting_iterator.h"
+#include "thrust/iterator/transform_iterator.h"
+
+namespace cudf {
+namespace detail {
+namespace {
+struct explode_functor {
+  /**
+   * @brief Function object for exploding a column.
+   */
+  template <typename T>
+  std::enable_if_t<!std::is_same<T, cudf::list_view>::value, std::unique_ptr<table>> operator()(
+    table_view const& input_table,
+    int const explode_column_idx,
+    rmm::cuda_stream_view stream,
+    rmm::mr::device_memory_resource* mr) const
+  {
+    CUDF_FAIL("Unsupported non-list column");
+  }
+
+  /**
+   * @brief Function object for exploding a column.
+   */
+   template <typename T>
+  std::enable_if_t<std::is_same<T, cudf::list_view>::value, std::unique_ptr<table>> operator()(
+    table_view const& input_table,
+    int const explode_column_idx,
+    rmm::cuda_stream_view stream,
+    rmm::mr::device_memory_resource* mr) const
+  {
```

Comment:
This would be cleaner to just use explicit specialization, e.g.,
```
template <typename T>
auto operator()(...){ // CUDF_FAIL }

template<>
auto operator()<list_view>(...){ // do stuff}
```
```